src/Processors/Executors/CompletedPipelineExecutor.cpp

```diff
@@ -72,9 +72,11 @@ void CompletedPipelineExecutor::execute()
         data->executor = std::make_shared<PipelineExecutor>(pipeline.processors, pipeline.process_list_element);
         data->executor->setReadProgressCallback(pipeline.getReadProgressCallback());
 
-        auto func = [&, thread_group = CurrentThread::getGroup()]()
+        /// Avoid passing this to labmda, copy ptr to data instead.
+        /// Destructor of unique_ptr copy raw ptr into local variable first, only then calls object destructor.
+        auto func = [data_ptr = data.get(), num_threads = pipeline.getNumThreads(), thread_group = CurrentThread::getGroup()]()
         {
-            threadFunction(*data, thread_group, pipeline.getNumThreads());
+            threadFunction(*data_ptr, thread_group, num_threads);
```

Comment:
Was able to reproduce by adding `std::this_thread::sleep_for(std::chrono::milliseconds(100));` here with
```
$ for i in $(seq 1 100); do timeout -s 2 0.031 clickhouse client --interactive_delay 1000 -q "insert into tab select number, toString(number) from system.numbers"; done
```

Won't add a test cause don't like to add another one setting for sleep.